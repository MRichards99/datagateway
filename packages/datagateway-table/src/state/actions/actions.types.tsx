import { Filter, Order, Entity } from 'datagateway-common';

// parent app actions
export const RegisterRouteType = 'daaas:api:register_route';
export const RequestPluginRerenderType = 'daaas:api:plugin_rerender';

// internal actions
export const SortTableType = 'datagateway_table:sort_table';
export const FilterTableType = 'datagateway_table:filter_table';

export const FetchInvestigationsRequestType =
  'datagateway_table:fetch_investigations_request';
export const FetchInvestigationsFailureType =
  'datagateway_table:fetch_investigations_failure';
export const FetchInvestigationsSuccessType =
  'datagateway_table:fetch_investigations_success';

export const FetchInvestigationCountRequestType =
  'datagateway_table:fetch_investigation_count_request';
export const FetchInvestigationCountFailureType =
  'datagateway_table:fetch_investigation_count_failure';
export const FetchInvestigationCountSuccessType =
  'datagateway_table:fetch_investigation_count_success';

export const FetchInvestigationDatasetsCountRequestType =
  'datagateway_table:fetch_investigation_datasets_count_request';
export const FetchInvestigationDatasetsCountFailureType =
  'datagateway_table:fetch_investigation_datasets_count_failure';
export const FetchInvestigationDatasetsCountSuccessType =
  'datagateway_table:fetch_investigation_datasets_count_success';

export const FetchDatasetsRequestType =
  'datagateway_table:fetch_datasets_request';
export const FetchDatasetsFailureType =
  'datagateway_table:fetch_datasets_failure';
export const FetchDatasetsSuccessType =
  'datagateway_table:fetch_datasets_success';

export const FetchDatasetCountRequestType =
  'datagateway_table:fetch_dataset_count_request';
export const FetchDatasetCountFailureType =
  'datagateway_table:fetch_dataset_count_failure';
export const FetchDatasetCountSuccessType =
  'datagateway_table:fetch_dataset_count_success';

export const FetchDatasetDatafilesCountRequestType =
  'datagateway_table:fetch_dataset_datafiles_count_request';
export const FetchDatasetDatafilesCountFailureType =
  'datagateway_table:fetch_dataset_datafiles_count_failure';
export const FetchDatasetDatafilesCountSuccessType =
  'datagateway_table:fetch_dataset_datafiles_count_success';

export const DownloadDatasetRequestType =
  'datagateway_table:download_dataset_request';
export const DownloadDatasetFailureType =
  'datagateway_table:download_dataset_failure';
export const DownloadDatasetSuccessType =
  'datagateway_table:download_dataset_success';

export const FetchDatafilesRequestType =
  'datagateway_table:fetch_datafiles_request';
export const FetchDatafilesFailureType =
  'datagateway_table:fetch_datafiles_failure';
export const FetchDatafilesSuccessType =
  'datagateway_table:fetch_datafiles_success';

export const FetchDatafileCountRequestType =
  'datagateway_table:fetch_datafile_count_request';
export const FetchDatafileCountFailureType =
  'datagateway_table:fetch_datafile_count_failure';
export const FetchDatafileCountSuccessType =
  'datagateway_table:fetch_datafile_count_success';

export const DownloadDatafileRequestType =
  'datagateway_table:download_datafile_request';
export const DownloadDatafileFailureType =
  'datagateway_table:download_datafile_failure';
export const DownloadDatafileSuccessType =
  'datagateway_table:download_datafile_success';

export const FetchInstrumentsRequestType =
  'datagateway_table:fetch_instruments_request';
export const FetchInstrumentsFailureType =
  'datagateway_table:fetch_instruments_failure';
export const FetchInstrumentsSuccessType =
  'datagateway_table:fetch_instruments_success';

export const FetchInstrumentCountRequestType =
  'datagateway_table:fetch_instrument_count_request';
export const FetchInstrumentCountFailureType =
  'datagateway_table:fetch_instrument_count_failure';
export const FetchInstrumentCountSuccessType =
  'datagateway_table:fetch_instrument_count_success';

export const FetchFacilityCyclesRequestType =
  'datagateway_table:fetch_facility_cycles_request';
export const FetchFacilityCyclesFailureType =
  'datagateway_table:fetch_facility_cycles_failure';
export const FetchFacilityCyclesSuccessType =
  'datagateway_table:fetch_facility_cycles_success';

export const FetchFacilityCycleCountRequestType =
  'datagateway_table:fetch_facility_cycle_count_request';
export const FetchFacilityCycleCountFailureType =
  'datagateway_table:fetch_facility_cycle_count_failure';
export const FetchFacilityCycleCountSuccessType =
  'datagateway_table:fetch_facility_cycle_count_success';

export interface SortTablePayload {
  column: string;
  order: Order | null;
}

export interface FilterTablePayload {
  column: string;
  filter: Filter | null;
}

export interface FailurePayload {
  error: string;
}

export interface FetchDataSuccessPayload {
  data: Entity[];
}

export interface FetchCountSuccessPayload {
  count: number;
}

export interface FetchDataCountSuccessPayload {
  id: number;
  count: number;
}
