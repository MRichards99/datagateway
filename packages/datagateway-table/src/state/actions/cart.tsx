import {
  FailurePayload,
  DownloadCartPayload,
  FetchDownloadCartSuccessType,
  FetchDownloadCartFailureType,
  FetchDownloadCartRequestType,
  AddToCartSuccessType,
  AddToCartFailureType,
  AddToCartRequestType,
  RemoveFromCartSuccessType,
  RemoveFromCartFailureType,
  RemoveFromCartRequestType,
} from './actions.types';
import { ActionType, ThunkResult } from '../app.types';
import { Action } from 'redux';
import axios from 'axios';
import * as log from 'loglevel';
import { DownloadCart } from 'datagateway-common';

export const fetchDownloadCartSuccess = (
  downloadCart: DownloadCart
): ActionType<DownloadCartPayload> => ({
  type: FetchDownloadCartSuccessType,
  payload: {
    downloadCart,
  },
});

export const fetchDownloadCartFailure = (
  error: string
): ActionType<FailurePayload> => ({
  type: FetchDownloadCartFailureType,
  payload: {
    error,
  },
});

export const fetchDownloadCartRequest = (): Action => ({
  type: FetchDownloadCartRequestType,
});

export const fetchDownloadCart = (): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    dispatch(fetchDownloadCartRequest());

    const { downloadApiUrl } = getState().dgtable.urls;

    // TODO: get facility name from somewhere else...
    await axios
      .get(`${downloadApiUrl}/user/cart/LILS`, {
        params: {
          // TODO: get session ID from somewhere else (extract from JWT)
          sessionId: window.localStorage.getItem('icat:token'),
        },
      })
      .then(response => {
        dispatch(fetchDownloadCartSuccess(response.data));
      })
      .catch(error => {
        log.error(error.message);
        dispatch(fetchDownloadCartFailure(error.message));
      });
  };
};

export const addToCartSuccess = (
  downloadCart: DownloadCart
): ActionType<DownloadCartPayload> => ({
  type: AddToCartSuccessType,
  payload: {
    downloadCart,
  },
});

export const addToCartFailure = (
  error: string
): ActionType<FailurePayload> => ({
  type: AddToCartFailureType,
  payload: {
    error,
  },
});

export const addToCartRequest = (): Action => ({
  type: AddToCartRequestType,
});

export const addToCart = (
  entityType: 'investigation' | 'dataset' | 'datafile',
  entityIds: number[]
): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    dispatch(addToCartRequest());

    const { downloadApiUrl } = getState().dgtable.urls;

    const params = new URLSearchParams();
    // TODO: get session ID from somewhere else (extract from JWT)
    params.append('sessionId', window.localStorage.getItem('icat:token') || '');
    params.append(
      'items',
      `${entityType} ${entityIds.join(`, ${entityType} `)}`
    );

    // TODO: get facility name from somewhere else...
    await axios
      .post(`${downloadApiUrl}/user/cart/LILS/cartItems`, params)
      .then(response => {
        dispatch(addToCartSuccess(response.data));
      })
      .catch(error => {
        log.error(error.message);
        dispatch(addToCartFailure(error.message));
      });
  };
};

export const removeFromCartSuccess = (
  downloadCart: DownloadCart
): ActionType<DownloadCartPayload> => ({
  type: RemoveFromCartSuccessType,
  payload: {
    downloadCart,
  },
});

export const removeFromCartFailure = (
  error: string
): ActionType<FailurePayload> => ({
  type: RemoveFromCartFailureType,
  payload: {
    error,
  },
});

export const removeFromCartRequest = (): Action => ({
  type: RemoveFromCartRequestType,
});

export const removeFromCart = (
  entityType: 'investigation' | 'dataset' | 'datafile',
  entityIds: number[]
): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    dispatch(removeFromCartRequest());

    const { downloadApiUrl } = getState().dgtable.urls;

    // TODO: get facility name from somewhere else...
    await axios
      .delete(`${downloadApiUrl}/user/cart/LILS/cartItems`, {
        params: {
          // TODO: get session ID from somewhere else (extract from JWT)
          sessionId: window.localStorage.getItem('icat:token'),
          items: `${entityType} ${entityIds.join(`, ${entityType} `)}`,
        },
      })
      .then(response => {
        dispatch(removeFromCartSuccess(response.data));
      })
      .catch(error => {
        log.error(error.message);
        dispatch(removeFromCartFailure(error.message));
      });
  };
};
