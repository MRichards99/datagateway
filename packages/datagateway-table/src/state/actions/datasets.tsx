import {
  FetchDatasetsSuccessType,
  FetchDatasetsFailureType,
  FetchDatasetsRequestType,
  FetchDatasetCountRequestType,
  FetchDatasetCountFailureType,
  FetchDatasetCountSuccessType,
  DownloadDatasetSuccessType,
  DownloadDatasetFailureType,
  DownloadDatasetRequestType,
  FetchDataSuccessPayload,
  FailurePayload,
  FetchDataCountSuccessPayload,
} from './actions.types';
import { ActionType, ThunkResult } from '../app.types';
import { source } from '../middleware/dgtable.middleware';
import { Action } from 'redux';
import axios from 'axios';
import { getApiFilter } from '.';
import { fetchDatafileCount } from './datafiles';
import * as log from 'loglevel';
import { Dataset } from 'datagateway-common';

export const fetchDatasetsSuccess = (
  datasets: Dataset[]
): ActionType<FetchDataSuccessPayload> => ({
  type: FetchDatasetsSuccessType,
  payload: {
    data: datasets,
  },
});

export const fetchDatasetsFailure = (
  error: string
): ActionType<FailurePayload> => ({
  type: FetchDatasetsFailureType,
  payload: {
    error,
  },
});

export const fetchDatasetsRequest = (): Action => ({
  type: FetchDatasetsRequestType,
});

export const fetchDatasets = (
  investigationId: number
): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    dispatch(fetchDatasetsRequest());

    let params = getApiFilter(getState);
    params.append(
      'where',
      JSON.stringify({ INVESTIGATION_ID: { eq: investigationId } })
    );
    const { datasetGetCount } = getState().dgtable.features;
    const { apiUrl } = getState().dgtable.urls;

    await axios
      .get(`${apiUrl}/datasets`, {
        params,
        headers: {
          Authorization: `Bearer ${window.localStorage.getItem('daaas:token')}`,
        },
      })
      .then(response => {
        dispatch(fetchDatasetsSuccess(response.data));
        if (datasetGetCount) {
          response.data.forEach((dataset: Dataset) => {
            dispatch(fetchDatafileCount(dataset.ID));
          });
        }
      })
      .catch(error => {
        log.error(error.message);
        dispatch(fetchDatasetsFailure(error.message));
      });
  };
};

export const downloadDatasetSuccess = (): Action => ({
  type: DownloadDatasetSuccessType,
});

export const downloadDatasetFailure = (
  error: string
): ActionType<FailurePayload> => ({
  type: DownloadDatasetFailureType,
  payload: {
    error,
  },
});

export const downloadDatasetRequest = (): Action => ({
  type: DownloadDatasetRequestType,
});

export const downloadDataset = (
  datasetId: number,
  datasetName: string
): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    dispatch(downloadDatasetRequest());

    const { idsUrl } = getState().dgtable.urls;

    // TODO: get ICAT session id properly when auth is sorted
    const params = {
      sessionId: window.localStorage.getItem('icat:token'),
      datasetIds: datasetId,
      compress: false,
      zip: true,
      outname: datasetName,
    };

    const link = document.createElement('a');
    link.href = `${idsUrl}/getData?${Object.entries(params)
      .map(([key, value]) => `${key}=${value}`)
      .join('&')}`;

    link.style.display = 'none';
    link.target = '_blank';
    document.body.appendChild(link);
    link.click();
    link.remove();
  };
};

export const fetchDatasetCountSuccess = (
  investigationId: number,
  count: number
): ActionType<FetchDataCountSuccessPayload> => ({
  type: FetchDatasetCountSuccessType,
  payload: {
    id: investigationId,
    count,
  },
});

export const fetchDatasetCountFailure = (
  error: string
): ActionType<FailurePayload> => ({
  type: FetchDatasetCountFailureType,
  payload: {
    error,
  },
});

export const fetchDatasetCountRequest = (): Action => ({
  type: FetchDatasetCountRequestType,
});

export const fetchDatasetCount = (
  investigationId: number
): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    dispatch(fetchDatasetCountRequest());

    const params = {
      where: {
        INVESTIGATION_ID: { eq: investigationId },
      },
    };
    const { apiUrl } = getState().dgtable.urls;

    // Check to see if a cached value exists already.
    const { investigationCache } = getState().dgtable;
    const datasetCount = investigationCache[investigationId];
    if (datasetCount) {
      console.log(
        'Cached dataset count value exists for investigation ID: ' +
          investigationId +
          ': ' +
          datasetCount
      );

      // Update dataset count with cached value
      // DGTableState.data.DATASET_COUNT = datasetCount
    } else {
      console.log(
        'Cached dataset count value does not exist for investigation ID (fetch from API): ' +
          investigationId
      );

      await axios
        .get(`${apiUrl}/datasets/count`, {
          params,
          headers: {
            Authorization: `Bearer ${window.localStorage.getItem(
              'daaas:token'
            )}`,
          },
          cancelToken: source.token,
        })
        .then(response => {
          dispatch(fetchDatasetCountSuccess(investigationId, response.data));
        })
        .catch(error => {
          log.error(error.message);
          dispatch(fetchDatasetCountFailure(error.message));
        });
    }
  };
};
