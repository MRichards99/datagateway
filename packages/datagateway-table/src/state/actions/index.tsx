import {
  ActionType,
  StateType,
  ThunkResult,
  ApplicationStrings,
} from '../app.types';
import {
  SortTablePayload,
  SortTableType,
  FilterTablePayload,
  FilterTableType,
  ClearTableType,
  ConfigureStringsType,
  ConfigureStringsPayload,
  FeatureSwitches,
  FeatureSwitchesPayload,
  ConfigureFeatureSwitchesType,
  URLs,
  ConfigureUrlsPayload,
  ConfigureURLsType,
  BreadcrumbSettings,
  ConfigureBreadcrumbSettingsPayload,
  ConfigureBreadcrumbSettingsType,
  SettingsLoadedType,
  ConfigureFacilityNamePayload,
  ConfigureFacilityNameType,
} from './actions.types';
import { Filter, Order } from 'datagateway-common';
import { Action } from 'redux';
import axios from 'axios';
import * as log from 'loglevel';
import { fetchDownloadCart } from './cart';

export const getApiFilter = (getState: () => StateType): URLSearchParams => {
  const sort = getState().dgtable.sort;
  const filters = getState().dgtable.filters;

  let searchParams = new URLSearchParams();

  for (let [key, value] of Object.entries(sort)) {
    searchParams.append('order', JSON.stringify(`${key} ${value}`));
  }

  // sort by ID first to guarantee order
  searchParams.append('order', JSON.stringify(`ID asc`));

  for (let [column, filter] of Object.entries(filters)) {
    if (typeof filter === 'object') {
      if ('startDate' in filter && filter.startDate) {
        searchParams.append(
          'where',
          JSON.stringify({ [column]: { gte: `${filter.startDate} 00:00:00` } })
        );
      }
      if ('endDate' in filter && filter.endDate) {
        searchParams.append(
          'where',
          JSON.stringify({ [column]: { lte: `${filter.endDate} 23:59:59` } })
        );
      }
    } else {
      searchParams.append(
        'where',
        JSON.stringify({ [column]: { like: filter } })
      );
    }
  }

  return searchParams;
};

export * from './investigations';
export * from './datasets';
export * from './datafiles';
export * from './instruments';
export * from './facilityCycles';
export * from './cart';

export const sortTable = (
  column: string,
  order: Order | null
): ActionType<SortTablePayload> => ({
  type: SortTableType,
  payload: {
    column,
    order,
  },
});

export const filterTable = (
  column: string,
  filter: Filter | null
): ActionType<FilterTablePayload> => ({
  type: FilterTableType,
  payload: {
    column,
    filter,
  },
});

export const clearTable = (): Action => ({
  type: ClearTableType,
});

export const settingsLoaded = (): Action => ({
  type: SettingsLoadedType,
});

export const configureStrings = (
  appStrings: ApplicationStrings
): ActionType<ConfigureStringsPayload> => ({
  type: ConfigureStringsType,
  payload: {
    res: appStrings,
  },
});

export const loadStrings = (path: string): ThunkResult<Promise<void>> => {
  return async dispatch => {
    await axios
      .get(path)
      .then(res => {
        dispatch(configureStrings(res.data));
      })
      .catch(error =>
        log.error(`Failed to read strings from ${path}: ${error}`)
      );
  };
};

export const loadFacilityName = (
  name: string
): ActionType<ConfigureFacilityNamePayload> => ({
  type: ConfigureFacilityNameType,
  payload: {
    facilityName: name,
  },
});

export const loadFeatureSwitches = (
  featureSwitches: FeatureSwitches
): ActionType<FeatureSwitchesPayload> => ({
  type: ConfigureFeatureSwitchesType,
  payload: {
    switches: featureSwitches,
  },
});

export const loadUrls = (urls: URLs): ActionType<ConfigureUrlsPayload> => ({
  type: ConfigureURLsType,
  payload: {
    urls,
  },
});

export const loadBreadcrumbSettings = (
  breadcrumbSettings: BreadcrumbSettings
): ActionType<ConfigureBreadcrumbSettingsPayload> => ({
  type: ConfigureBreadcrumbSettingsType,
  payload: {
    settings: breadcrumbSettings,
  },
});

export const configureApp = (): ThunkResult<Promise<void>> => {
  return async dispatch => {
    await axios
      .get('/datagateway-table-settings.json')
      .then(res => {
        const settings = res.data;

        // invalid settings.json
        if (typeof settings !== 'object') {
          throw Error('Invalid format');
        }

        // Get the facility name from settings.
        if ('facilityName' in settings) {
          dispatch(loadFacilityName(settings['facilityName']));
        } else {
          throw new Error('facilityName is undefined in settings');
        }

        // features is an optional setting
        if ('features' in settings) {
          dispatch(loadFeatureSwitches(settings['features']));
        }

        if (
          'idsUrl' in settings &&
          'apiUrl' in settings &&
          'downloadApiUrl' in settings
        ) {
          dispatch(
            loadUrls({
              idsUrl: settings['idsUrl'],
              apiUrl: settings['apiUrl'],
              downloadApiUrl: settings['downloadApiUrl'],
            })
          );
        } else {
          throw new Error(
            'One of the URL options (idsUrl, apiUrl, downloadApiUrl) is undefined in settings'
          );
        }

        // Dispatch the action to load the breadcrumb settings (optional settings).
        if ('breadcrumbs' in settings) {
          dispatch(loadBreadcrumbSettings(settings['breadcrumbs']));
        }

        /* istanbul ignore if */
        if (process.env.NODE_ENV === `development`) {
          // TODO: get info from correct places when authorisation is sorted out in parent app
          axios
            .post(`${settings['apiUrl']}/sessions`, {
              username: 'user',
              password: 'password',
            })
            .then(response => {
              window.localStorage.setItem(
                'daaas:token',
                response.data.sessionID
              );
            })
            .catch(error => {
              log.error(`Can't contact API: ${error.message}`);
            });

          // TODO: replace with getting from daaas:token when supported
          const splitUrl = settings.downloadApiUrl.split('/');
          const icatUrl = `${splitUrl
            .slice(0, splitUrl.length - 1)
            .join('/')}/icat`;
          axios
            .post(
              `${icatUrl}/session`,
              `json=${JSON.stringify({
                plugin: 'simple',
                credentials: [{ username: 'root' }, { password: 'pw' }],
              })}`,
              {
                headers: {
                  'Content-Type': 'application/x-www-form-urlencoded',
                },
              }
            )
            .then(response => {
              window.localStorage.setItem(
                'icat:token',
                response.data.sessionId
              );
            })
            .catch(error => log.error("Can't log in to ICAT"));
        }

        if ('ui-strings' in settings) {
          const uiStringResourcesPath = !settings['ui-strings'].startsWith('/')
            ? '/' + settings['ui-strings']
            : settings['ui-strings'];
          dispatch(loadStrings(uiStringResourcesPath));
        }

        // fetch initial download cart
        dispatch(fetchDownloadCart());

        dispatch(settingsLoaded());
      })
      .catch(error => {
        log.error(
          `Error loading datagateway-table-settings.json: ${error.message}`
        );
      });
  };
};
