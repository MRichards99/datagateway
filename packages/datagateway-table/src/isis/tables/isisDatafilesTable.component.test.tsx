import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import ISISDatafilesTable from './isisDatafilesTable.component';
import { initialState as dgTableInitialState } from '../../state/reducers/dgtable.reducer';
import configureStore from 'redux-mock-store';
import { StateType } from '../../state/app.types';
import {
  fetchDatafilesRequest,
  filterTable,
  sortTable,
  downloadDatafileRequest,
  fetchDatafileDetailsRequest,
  fetchDatafileCountRequest,
  removeFromCartRequest,
  addToCartRequest,
  dGCommonInitialState,
} from 'datagateway-common';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { Table, Datafile } from 'datagateway-common';
import { MemoryRouter } from 'react-router';
import axios from 'axios';
import { clearTable } from '../../state/actions';

describe('ISIS datafiles table component', () => {
  let shallow;
  let mount;
  let mockStore;
  let state: StateType;
  (axios.get as jest.Mock).mockImplementation(() =>
    Promise.resolve({ data: [] })
  );
  global.Date.now = jest.fn(() => 1);

  beforeEach(() => {
    shallow = createShallow({ untilSelector: 'ISISDatafilesTable' });
    mount = createMount();

    mockStore = configureStore([thunk]);
    state = JSON.parse(
      JSON.stringify({
        dgtable: dgTableInitialState,
        dgcommon: dGCommonInitialState,
      })
    );
    state.dgcommon.data = [
      {
        ID: 1,
        NAME: 'Test 1',
        LOCATION: '/test1',
        FILESIZE: 1,
        CREATE_TIME: '2019-07-23',
        DATASET_ID: 1,
      },
    ];
    state.dgcommon.allIds = [1];
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('renders correctly', () => {
    const wrapper = shallow(
      <ISISDatafilesTable store={mockStore(state)} datasetId="1" />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('sends clearTable action on load', () => {
    const testStore = mockStore(state);
    mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <ISISDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(clearTable());
  });

  it('sends fetchDatafileCount and fetchDatafiles actions when watched store values change', () => {
    let testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <ISISDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    // simulate clearTable action
    testStore = mockStore({
      ...state,
      dgtable: { ...state.dgtable, sort: {}, filters: {} },
    });
    wrapper.setProps({ store: testStore });

    expect(testStore.getActions()[1]).toEqual(fetchDatafileCountRequest(1));
    expect(testStore.getActions()[2]).toEqual(fetchDatafilesRequest(1));
  });

  it('sends fetchDatafiles action when loadMoreRows is called', () => {
    const testStore = mockStore(state);
    const wrapper = shallow(
      <ISISDatafilesTable datasetId="1" store={testStore} />
    );

    wrapper.prop('loadMoreRows')({ startIndex: 50, stopIndex: 74 });

    expect(testStore.getActions()[0]).toEqual(fetchDatafilesRequest(1));
  });

  it('sends filterTable action on text filter', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <ISISDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    const filterInput = wrapper.find('[aria-label="Filter by Name"] input');
    filterInput.instance().value = 'test';
    filterInput.simulate('change');

    expect(testStore.getActions()[1]).toEqual(filterTable('NAME', 'test'));

    filterInput.instance().value = '';
    filterInput.simulate('change');

    expect(testStore.getActions()[2]).toEqual(filterTable('NAME', null));
  });

  it('sends filterTable action on date filter', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <ISISDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    const filterInput = wrapper.find(
      '[aria-label="Modified Time date filter to"]'
    );
    filterInput.instance().value = '2019-08-06';
    filterInput.simulate('change');

    expect(testStore.getActions()[1]).toEqual(
      filterTable('MOD_TIME', { endDate: '2019-08-06' })
    );

    filterInput.instance().value = '';
    filterInput.simulate('change');

    expect(testStore.getActions()[2]).toEqual(filterTable('MOD_TIME', null));
  });

  it('sends sortTable action on sort', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <ISISDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    wrapper
      .find('[role="columnheader"] span[role="button"]')
      .first()
      .simulate('click');

    expect(testStore.getActions()[1]).toEqual(sortTable('NAME', 'asc'));
  });

  it('sends addToCart action on unchecked checkbox click', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <ISISDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    wrapper
      .find('[aria-label="select row 0"]')
      .first()
      .simulate('click');

    expect(testStore.getActions()[1]).toEqual(addToCartRequest());
  });

  it('sends removeFromCart action on checked checkbox click', () => {
    state.dgtable.cartItems = [
      {
        entityId: 1,
        entityType: 'datafile',
        id: 1,
        name: 'test',
        parentEntities: [],
      },
    ];

    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <ISISDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    wrapper
      .find('[aria-label="select row 0"]')
      .first()
      .simulate('click');

    expect(testStore.getActions()[1]).toEqual(removeFromCartRequest());
  });

  it('selected rows only considers relevant cart items', () => {
    state.dgtable.cartItems = [
      {
        entityId: 1,
        entityType: 'dataset',
        id: 1,
        name: 'test',
        parentEntities: [],
      },
      {
        entityId: 2,
        entityType: 'datafile',
        id: 2,
        name: 'test',
        parentEntities: [],
      },
    ];

    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <ISISDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    const selectAllCheckbox = wrapper
      .find('[aria-label="select all rows"]')
      .first();

    expect(selectAllCheckbox.prop('checked')).toEqual(false);
    expect(selectAllCheckbox.prop('data-indeterminate')).toEqual(false);
  });

  it('sends downloadData action on click of download button', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <ISISDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    wrapper.find('button[aria-label="Download"]').simulate('click');

    expect(testStore.getActions()[1]).toEqual(downloadDatafileRequest(1));
  });

  it("doesn't display download button for datafiles with no location", () => {
    const datafile = state.dgtable.data[0] as Datafile;
    const { LOCATION, ...datafileWithoutLocation } = datafile;
    state.dgtable.data = [datafileWithoutLocation];

    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <ISISDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('button[aria-label="Download"]')).toHaveLength(0);
  });

  it('renders details panel correctly and it sends off an FetchDatasetDetails action', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <ISISDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    const detailsPanelWrapper = shallow(
      wrapper.find(Table).prop('detailsPanel')({
        rowData: state.dgtable.data[0],
      })
    );
    expect(detailsPanelWrapper).toMatchSnapshot();

    mount(
      wrapper.find(Table).prop('detailsPanel')({
        rowData: state.dgtable.data[0],
        detailsPanelResize: jest.fn(),
      })
    );

    expect(testStore.getActions()[1]).toEqual(fetchDatafileDetailsRequest());
  });

  it('renders file size as bytes', () => {
    const wrapper = mount(
      <Provider store={mockStore(state)}>
        <MemoryRouter>
          <ISISDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    expect(
      wrapper
        .find('[aria-colindex=5]')
        .find('p')
        .text()
    ).toEqual('1 B');
  });
});
