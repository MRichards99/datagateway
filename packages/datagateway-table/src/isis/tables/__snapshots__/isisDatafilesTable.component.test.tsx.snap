// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ISIS datafiles table component renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "height": "calc(100vh - 64px)",
      "width": "100%",
    }
  }
>
  <WithStyles(VirtualizedTable)
    actions={
      Array [
        [Function],
      ]
    }
    columns={
      Array [
        Object {
          "dataKey": "NAME",
          "filterComponent": [Function],
          "label": "Name",
        },
        Object {
          "dataKey": "LOCATION",
          "filterComponent": [Function],
          "label": "Location",
        },
        Object {
          "cellContentRenderer": [Function],
          "dataKey": "FILESIZE",
          "filterComponent": [Function],
          "label": "Size",
        },
        Object {
          "dataKey": "MOD_TIME",
          "filterComponent": [Function],
          "label": "Modified Time",
        },
      ]
    }
    data={
      Array [
        Object {
          "CREATE_TIME": "2019-07-23",
          "DATASET_ID": 1,
          "FILESIZE": 1,
          "ID": 1,
          "LOCATION": "/test1",
          "NAME": "Test 1",
        },
      ]
    }
    detailsPanel={[Function]}
    loadMoreRows={[Function]}
    onSort={[Function]}
    sort={Object {}}
    totalRowCount={0}
  />
</div>
`;

exports[`ISIS datafiles table component renders details panel correctly and it sends off an FetchDatasetDetails action 1`] = `
<div>
  <WithStyles(ForwardRef(Tabs))
    aria-label="datafile-details-tabs"
    onChange={[Function]}
    value="details"
  >
    <WithStyles(ForwardRef(Tab))
      aria-controls="datafile-details-panel"
      id="datafile-details-tab"
      label="Datafile Details"
      value="details"
    />
  </WithStyles(ForwardRef(Tabs))>
  <div
    aria-labelledby="datafile-details-tab"
    hidden={false}
    id="datafile-details-panel"
    role="tabpanel"
  >
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Name:
      </b>
       
      Test 1
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Description:
      </b>
       
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Location:
      </b>
       
      /test1
    </WithStyles(ForwardRef(Typography))>
  </div>
</div>
`;
